### DISCUSSION TITLE
What is the best way to compare vectors from imagenet? Is euclidean distance  a good choice?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [exadel-inc/CompreFace#819] [Q&A] [answered]
What is the best way to compare vectors from imagenet? 
Is Euclidean distance  a good choice? 

is there a better algorithm to find similar faces?

**Comment by pospielov** [comment, marked as answer]
we use Euclidean distance because this is what is used by default in the open-source facial recognition libraries we use. It guarantees that the accuracy is the same as in the original library.
There is another algorithm - Cosine similarity. Most libraries allow to use it as well. But it will make CompreFace more complex to implement and use. So we didn't add it.

**Comment by martinenkoEduard** [comment]
What are the best way to store vectors to milvus database? 
I get worse results in milvus, that in compreface itself.
But comrpeface can only work with database 200k faces, and milvus can store several millions..

Does it use Euclidean distance (L2)? Or something different?

How can I improve search results using milvus?
Is it better to use
Euclidean distance (L2)
or
Inner product (IP)
?

Maybe somehow change vectors(normalize or something else) before inserting to milvus?

What distance metric algorithm compreface uses to compare two faces?
Does it somehow changes vectors?

**Reply by pospielov** [reply]
CompreFace uses Euclidean distance (L2).
CompreFace normalizes vectors before calculating L2
You can see all transformations here:
https://github.com/exadel-inc/CompreFace/blob/master/java/api/src/main/java/com/exadel/frs/core/trainservice/component/classifiers/EuclideanDistanceClassifier.java

Could you describe more, how bad is the milvus results?
